File: pset6/credit.py
#Python implementation of the creadit card checker

import cs50

def main():
    while True:
        print("Credit card number:", end=" ")
        card = cs50.get_string()
        if int(card) >= 0:
            break

    number = [int(i) for i in card]
    sum = 0
    number.reverse()

    for i in range(len(number)):
        if i % 2 == 0:
            sum += number[i]
        else:
            if 2*number[i] < 9:
                sum += 2*number[i]
            else:
                sum += 2*number[i]//10 + 2*number[i]%10

    if sum % 10 == 0:
        if number[-1] == 3 and number[-2] == 4 and len(number) == 15: 
            print("AMEX")
        elif number[-1] == 3 and number[-2] == 7 and len(number) == 15: 
            print("AMEX")
        elif number[-1] == 5 and (number[-2] < 6 or number[-2] > 0) and len(number) == 16: 
            print("MASTERCARD")
        elif number[-1] == 4 and len(number) == 13: 
            print("VISA")
        elif number[-1] == 4 and len(number) == 16: 
            print("VISA")
        else: 
            print("INVALID")
    else: 
        print("INVALID")
    


if __name__ == "__main__":
    main()

File: hacker1/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    
  long long a;
  do
  {
    printf("Credit card number:");
    a = GetLongLong();
  }
  while (a <= 0);
    
  int number[16];
  int digits = 0;
  int sum = 0;
    
  while (a > 0) 
  {
    number[digits] = a % 10;
    a = a / 10;
    digits++;
  }
   
  digits--; 
    
  for (int i = 1; i <= digits; i = i + 2)
  {
    if (2*number[i] > 9)
      sum = sum + ((2 * number[i]) / 10) + ((2 * number[i]) % 10);
    else 
      sum = sum + (2 * number[i]); 
  } 
    
  for (int i = 0; i <= digits; i = i + 2)
    sum = sum + number[i];

  if (sum % 10 == 0)
  {
    if ((number[digits] == 3) && (number[digits-1] == 4) && digits == 14) 
      printf("AMEX\n");
    else if ((number[digits] == 3) && (number[digits-1] == 7) && digits == 14) 
      printf("AMEX\n");
    else if ((number[digits] == 5) && ((number[digits-1] < 6) || (number[digits-1] > 0))) 
      printf("MASTERCARD\n");
    else if (number[digits] == 4 && digits == 12) 
      printf("VISA\n");
    else if (number[digits] == 4 && digits == 15) 
      printf("VISA\n");
    else 
      printf("INVALID\n");
  }
    
  else 
    printf("INVALID\n");

  return 0;

}